execbuilder: limit assignment_cast function search to builtins

mw5h opened the pull request:
Previously execbuilder would search the function catalog for the assignment_cast function. This was inefficient and could possibly lead to incorrect behavior if an application defined a function called crdb_internal.assignment_cast. Instead of resolving the function in this manner, we now just consult the map of builtin functions.

Fixes: #121460
Releasenote: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126570)
<!-- Reviewable:end -->

DrewKimball reviewed (commented):
Nice fix! Is there a regression test we could add? Also, should we also change the [callsite](https://github.com/cockroachdb/cockroach/blob/38e6f37ff5d339d6205cf76b5bc2cc5febf361eb/pkg/sql/opt/exec/execbuilder/relational.go#L3645) for window functions as well?

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126570#-:-O0p8vAx1H33veJHabHB:b-z7tmde)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mw5h)



<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
Very nice!

nit: In the future you can add **cockroachdb/sql-queries** as a reviewer which notifies everyone on the team about the PR. The **cockroachdb/sql-queries-pr** reviewer is only used to assign a random reviewer.

![image.png](https://files.reviewableusercontent.io/images/reviewable/1128750/SuTdQCtkQangasSNnkVnAhMY/image.png)

> Nice fix! Is there a regression test we could add? ...

We could if we found a way to trigger a bug after adding a built-in function with the same qualified name as `crdb_internal.assignment_cast`. I'm not that such a bug exists or is visible, but might be worth a shot.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126570#-:-O0sn9TY8xuYTlrOhbgy:bik8jho)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`-- commits` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126570#-O0snWDwA_FCpIIsiuZG):*
nit: The format must be `Release note: None` for the release script to work.


<!-- Sent from Reviewable.io -->

mw5h commented:
Still looking at a regression test.

mw5h commented:
```
demo@127.0.0.1:26257/demoapp/movr> CREATE FUNCTION crdb_internal.assignment_cast() RETURNS integer AS $$
                                -> SELECT 42 AS result;
                                -> $$ LANGUAGE SQL;
ERROR: crdb_internal is a virtual schema and cannot be modified
SQLSTATE: 42501
```

DrewKimball reviewed (approved):
If we can't find a repro with reasonable effort, the fix alone <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126570#-:-O0taKU86g5YEWfnNLBp:b6fnsqw)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DarrylWong, @dt, @mw5h, and @vidit-bhat)



<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126570)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DarrylWong, @DrewKimball, @dt, @mw5h, and @vidit-bhat)

___
*[`-- commits` line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126570#-O0tv9VaE_ZHqXNZD1FE):*
nit: Looks like it's a bit more complicated than consulting the map directly, so you might want to update this sentence.
___
*[`pkg/sql/opt/exec/execbuilder/builder.go` line 276 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126570#-O0tvgev3yhbBK7OFeQ7:-O0tvgewCZ6hZtLBR--9:b-j55xb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a1ef66fc75fb60bc9187e9ca1dfbc185b0df7fa/pkg/sql/opt/exec/execbuilder/builder.go#L276)):*
> ```Go
> 			return nil, err
> 		}
> 		return tree.GetBuiltinFuncDefinitionOrFail(uname, path)
> ```

I kind of liked the first implementation where the map is consulted directly, because the search path should be irrelevant when resolving a known built-in function. Was there problems with that method?


<!-- Sent from Reviewable.io -->

mw5h commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126570)_ status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DarrylWong, @DrewKimball, @dt, @mw5h, and @vidit-bhat)
> 
> _[`-- commits` line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126570#-O0tv9VaE_ZHqXNZD1FE):_ nit: Looks like it's a bit more complicated than consulting the map directly, so you might want to update this sentence.
> 
> _[`pkg/sql/opt/exec/execbuilder/builder.go` line 276 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126570#-O0tvgev3yhbBK7OFeQ7:-O0tvgewCZ6hZtLBR--9:b-j55xb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a1ef66fc75fb60bc9187e9ca1dfbc185b0df7fa/pkg/sql/opt/exec/execbuilder/builder.go#L276)):_
> 
> > ```go
> > 			return nil, err
> > 		}
> > 		return tree.GetBuiltinFuncDefinitionOrFail(uname, path)
> > ```
> 
> I kind of liked the first implementation where the map is consulted directly, because the search path should be irrelevant when resolving a known built-in function. Was there problems with that method?

It doesn't find window functions. This patch is also broken and is failing some tests.

mw5h committed:
execbuilder: limit internal function search to builtins

Previously execbuilder would search the function catalog for the
assignment_cast and window functions. This was inefficient and could
possibly lead to incorrect behavior if an application defined a
function called an appropriately named function (e.g.
crdb_internal.assignment_cast). Instead of resolving the function
in this manner, we now limit our search to built-in functions.

Fixes: #121460
Release note: None

mw5h commented:
I've broken this patch into two sub-patches. The first addresses the original complaint with minimum complexity, the second addresses @DrewKimball's original comments at the expense of additional complexity. My intuition is that casting is quite a bit more common than window functions and more likely to occur on latency sensitive queries, making the simpler patch the better option here. Does that sound right?

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!

> Reviewed 2 of 3 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126570#-:-O1T2Rb38aD5qFWIp4ff:bcmw7ss)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DarrylWong, @dt, @mw5h, and @vidit-bhat)

___
*[`-- commits` line 6 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126570#-O1T1m620XhmWOOcRdiO):*
nit: resulve -> resolve
___
*[`pkg/sql/opt/exec/execbuilder/builder.go` line 276 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126570#-O0tvgev3yhbBK7OFeQ7:-O1T1xJkCU8G5Euwjm7c:bb73djm) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a1ef66fc75fb60bc9187e9ca1dfbc185b0df7fa/pkg/sql/opt/exec/execbuilder/builder.go#L276)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

It doesn't find window functions. This patch is also broken and is failing some tests.
</blockquote></details>

SGTM


<!-- Sent from Reviewable.io -->

mw5h commented:
bors r+
